File: src/observer.rs
//! An observer which can monitor a controller <-> gateway network.
//!
//! ```text
//! ┌───┐
//! │TAP│◁ ─ ─ ─ …
//! └───┘
//!   ▲
//!   │
//!   │
//!   ├──────┐
//!   ▼      ▼
//! ┌───┐  ┌───┐
//! │CCA│  │O_o│
//! └───┘  └───┘
//! ```

use crate::gateway::link::GatewayID;
use crate::pv::application::{NodeTableResponseEntry, TopologyReport};
use crate::pv::link::SlotCounter;
use crate::pv::network::{NodeAddress, ReceivedPacketHeader};
use crate::pv::{LongAddress, NodeID, PacketType};
use crate::{gateway, pv};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::collections::btree_map::Entry;
use std::collections::BTreeMap;
use std::time::SystemTime;

pub mod event;

mod persistent_state;
use persistent_state::{PersistentState, PersistentStateEvent};

mod node_table;
use node_table::NodeTableBuilder;

mod slot_clock;
use slot_clock::SlotClock;

/// An observer, monitoring a controller interacting with one or more TAPs via an RS-485 interface.
#[derive(Debug)]
pub struct Observer {
    persistent_state: PersistentState,

    enumeration_state: Option<EnumerationState>,
    captured_slot_counters: BTreeMap<GatewayID, SystemTime>,
    slot_clocks: BTreeMap<GatewayID, SlotClock>,
    node_table_builders: BTreeMap<GatewayID, NodeTableBuilder>,
}

impl Default for Observer {
    fn default() -> Self {
        Self::from_persistent_state(PersistentState::default())
    }
}

impl Observer {
    pub fn from_persistent_state(persistent_state: PersistentState) -> Self {
        Observer {
            persistent_state,
            enumeration_state: None,
            captured_slot_counters: Default::default(),
            slot_clocks: Default::default(),
            node_table_builders: Default::default(),
        }
    }

    pub fn persistent_state(&self) -> &PersistentState {
        &self.persistent_state
    }
}

impl gateway::transport::Sink for Observer {
    fn enumeration_started(&mut self, enumeration_gateway_id: GatewayID) {
        self.enumeration_state = Some(EnumerationState {
            enumeration_gateway_id,
            gateway_identities: Default::default(),
            gateway_versions: Default::default(),
        });
    }

    fn gateway_identity_observed(&mut self, gateway_id: GatewayID, address: LongAddress) {
        if let Some(enumeration_state) = self.enumeration_state.as_mut() {
            // We're enumerating
            // Delegate
            enumeration_state.gateway_identity_observed(gateway_id, address);
        } else {
            // Accept the identity as-is
            self.persistent_state
                .gateway_identities
                .insert(gateway_id, address);
            let infrastructure_event = PersistentStateEvent::from(&self.persistent_state);
            println!("{}", serde_json::to_string(&infrastructure_event).unwrap());
        }
    }

    fn gateway_version_observed(&mut self, gateway_id: GatewayID, version: &str) {
        let version = version.to_owned();

        if let Some(enumeration_state) = self.enumeration_state.as_mut() {
            enumeration_state
                .gateway_versions
                .insert(gateway_id, version);
        } else {
            self.persistent_state
                .gateway_versions
                .insert(gateway_id, version);
            let infrastructure_event = PersistentStateEvent::from(&self.persistent_state);
            println!("{}", serde_json::to_string(&infrastructure_event).unwrap());
        }
    }

    fn enumeration_ended(&mut self, _gateway_id: GatewayID) {
        // We're done enumerating
        // Did we catch the whole exchange?
        if let Some(enumeration_state) = self.enumeration_state.take() {
            // Accept the gateway information learned during enumeration as a replacement for our
            // existing state
            self.persistent_state.gateway_identities = enumeration_state.gateway_identities;
            self.persistent_state.gateway_versions = enumeration_state.gateway_versions;
            let infrastructure_event = PersistentStateEvent::from(&self.persistent_state);
            println!("{}", serde_json::to_string(&infrastructure_event).unwrap());
        }
    }

    fn gateway_slot_counter_captured(&mut self, gateway_id: GatewayID) {
        self.captured_slot_counters
            .insert(gateway_id, SystemTime::now());
    }

    fn gateway_slot_counter_observed(&mut self, gateway_id: GatewayID, slot_counter: SlotCounter) {
        let Some(time) = self.captured_slot_counters.remove(&gateway_id) else {
            return;
        };

        match self.slot_clocks.entry(gateway_id) {
            Entry::Vacant(e) => {
                if let Ok(clock) = SlotClock::new(slot_counter, time) {
                    e.insert(clock);
                }
            }
            Entry::Occupied(mut e) => {
                e.get_mut().set(slot_counter, time).ok();
            }
        }
    }

    fn packet_received(
        &mut self,
        _gateway_id: GatewayID,
        _packet_header: &ReceivedPacketHeader,
        _packet_data: &[u8],
    ) {
    }

    fn command_executed(
        &mut self,
        _gateway_id: GatewayID,
        _command_request: (PacketType, &[u8]),
        _command_response: (PacketType, &[u8]),
    ) {
    }
}

impl pv::application::Sink for Observer {
    fn string_request(
        &mut self,
        _gateway_id: GatewayID,
        _pv_node_id: NodeID,
        _string_request: &str,
    ) {
    }

    fn string_response(
        &mut self,
        _gateway_id: GatewayID,
        _pv_node_id: NodeID,
        _string_response: &str,
    ) {
    }

    fn node_table_page(
        &mut self,
        gateway_id: GatewayID,
        start_address: NodeAddress,
        nodes: &[NodeTableResponseEntry],
    ) {
        let builder = self.node_table_builders.entry(gateway_id).or_default();

        if let Some(new_table) = builder.push(start_address, nodes) {
            self.persistent_state
                .gateway_node_tables
                .insert(gateway_id, new_table);
            let infrastructure_event = PersistentStateEvent::from(&self.persistent_state);
            println!("{}", serde_json::to_string(&infrastructure_event).unwrap());
        }
    }

    fn topology_report(
        &mut self,
        _gateway_id: GatewayID,
        _pv_node_id: NodeID,
        _topology_report: &TopologyReport,
    ) {
    }

    fn power_report(
        &mut self,
        gateway_id: GatewayID,
        node_id: NodeID,
        power_report: &pv::application::PowerReport,
    ) {
        let Some(slot_clock) = self.slot_clocks.get(&gateway_id) else {
            log::error!(
                "discarding power report from gateway {:?} due to missing slot clock: {:?}",
                gateway_id,
                power_report
            );
            return;
        };

        let Ok(event) = event::PowerReportEvent::new(gateway_id, node_id, slot_clock, power_report)
        else {
            log::error!(
                "discarding power report from gateway {:?} due to invalid slot counter: {:?}",
                gateway_id,
                power_report
            );
            return;
        };

        println!("{}", serde_json::to_string(&event).unwrap());
    }
}

#[derive(Debug, Clone, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
struct EnumerationState {
    enumeration_gateway_id: GatewayID,
    gateway_identities: BTreeMap<GatewayID, LongAddress>,
    gateway_versions: BTreeMap<GatewayID, String>,
}

impl EnumerationState {
    fn gateway_identity_observed(&mut self, gateway: GatewayID, address: LongAddress) {
        // Is this a persistent ID?
        if gateway == self.enumeration_gateway_id {
            // No, it's the enumeration address
            // Discard this response, since we'll get a persistent one shortly
            return;
        }

        // Store the identity
        self.gateway_identities.insert(gateway, address);
    }
}

#[cfg(test)]
mod tests;

File: python/taptap/observer/observer.py
"""Main Observer class for monitoring TAP protocol."""

import json
from pathlib import Path
from typing import Optional, Dict, List, Tuple
from datetime import datetime
import sys

from ..gateway.link import GatewayID
from ..pv.link import SlotCounter
from ..pv.network import NodeID, NodeAddress
from ..pv.network.types import LongAddress
from ..pv.application import PowerReport
from .slot_clock import SlotClock
from .event import PowerReportEvent, Event
from .persistent_state import PersistentState
from .node_table import NodeTable


class _EnumerationState:
    """Tracks an active gateway enumeration cycle.
    
    During enumeration, gateway identities and versions are collected
    and applied atomically when enumeration ends.
    """
    
    def __init__(self, enumeration_gateway_id: GatewayID):
        self.enumeration_gateway_id = enumeration_gateway_id
        self.gateway_identities: Dict[int, LongAddress] = {}
        self.gateway_versions: Dict[int, str] = {}
    
    def gateway_identity_observed(self, gateway_id: GatewayID, address: LongAddress) -> None:
        """Record a gateway identity during enumeration."""
        # Filter out the temporary enumeration address
        if gateway_id.value == self.enumeration_gateway_id.value:
            return
        self.gateway_identities[gateway_id.value] = address
    
    def gateway_version_observed(self, gateway_id: GatewayID, version: str) -> None:
        """Record a gateway version during enumeration."""
        if gateway_id.value == self.enumeration_gateway_id.value:
            return
        self.gateway_versions[gateway_id.value] = version


class Observer:
    """Observer monitors a controller interacting with TAP gateways.
    
    Implements both transport.Sink and application.Sink to receive
    all protocol events and generate JSON output.
    """
    
    def __init__(self, state_file: Optional[Path] = None):
        """Create observer with optional state file.
        
        Args:
            state_file: Path to JSON file for persistent state (optional)
        """
        self._state_file = state_file
        self._persistent_state = PersistentState()
        self._enumeration_state: Optional[_EnumerationState] = None
        self._slot_clocks: Dict[int, SlotClock] = {}  # gateway_id.value -> SlotClock
        self._captured_slot_counters: Dict[int, datetime] = {}  # gateway_id.value -> time
        self._node_table_builders: Dict[int, Dict[int, LongAddress]] = {}  # gw -> partial table
        
        # Read existing state if file exists
        self._read_persistent_state()
    
    def _read_persistent_state(self) -> None:
        """Load persistent state from JSON file if it exists."""
        if self._state_file is None:
            return
        
        if not self._state_file.exists():
            return
        
        try:
            with open(self._state_file, 'r') as f:
                data = json.load(f)
                self._persistent_state = PersistentState.from_dict(data)
            
            # Emit infrastructure event
            event = self._persistent_state.to_infrastructure_event()
            print(json.dumps(event), flush=True)
            
        except Exception as e:
            print(f"Warning: Failed to read persistent state: {e}", file=sys.stderr)
    
    def write_persistent_state(self) -> None:
        """Write persistent state to JSON file atomically."""
        if self._state_file is None:
            return
        
        try:
            # Write to temporary file first
            tmp_path = self._state_file.with_suffix('.tmp')
            with open(tmp_path, 'w') as f:
                json.dump(self._persistent_state.to_dict(), f, indent=2)
            
            # Atomic rename
            tmp_path.replace(self._state_file)
            
            # Emit infrastructure event
            event = self._persistent_state.to_infrastructure_event()
            print(json.dumps(event), flush=True)
            
        except Exception as e:
            print(f"Error: Failed to write persistent state: {e}", file=sys.stderr)
    
    # Implement transport.Sink — enumeration methods
    
    def enumeration_started(self, enumeration_gateway_id: GatewayID) -> None:
        """Enumeration started, using the indicated gateway ID.
        
        Args:
            enumeration_gateway_id: Temporary gateway ID used during enumeration
        """
        self._enumeration_state = _EnumerationState(enumeration_gateway_id)
    
    def gateway_identity_observed(self, gateway_id: GatewayID, address: LongAddress) -> None:
        """A gateway's hardware address was observed.
        
        Args:
            gateway_id: Gateway ID (may be temporary during enumeration)
            address: Gateway's 8-byte hardware address
        """
        if self._enumeration_state is not None:
            self._enumeration_state.gateway_identity_observed(gateway_id, address)
        else:
            self._persistent_state.gateway_identities[gateway_id.value] = address
            self.write_persistent_state()
    
    def gateway_version_observed(self, gateway_id: GatewayID, version: str) -> None:
        """A gateway's version string was observed.
        
        Args:
            gateway_id: Gateway ID
            version: Version string
        """
        if self._enumeration_state is not None:
            self._enumeration_state.gateway_version_observed(gateway_id, version)
        else:
            self._persistent_state.gateway_versions[gateway_id.value] = version
            self.write_persistent_state()
    
    def enumeration_ended(self, gateway_id: GatewayID) -> None:
        """Enumeration ended — apply collected identities atomically.
        
        Args:
            gateway_id: Gateway ID that ended enumeration
        """
        if self._enumeration_state is not None:
            # Replace persistent state with enumeration results
            self._persistent_state.gateway_identities = self._enumeration_state.gateway_identities
            self._persistent_state.gateway_versions = self._enumeration_state.gateway_versions
            self._enumeration_state = None
            self.write_persistent_state()
    
    # Implement transport.Sink — slot counter methods
    
    def gateway_slot_counter_captured(self, gateway_id: GatewayID) -> None:
        """Gateway captured its slot counter (in RECEIVE_REQUEST).
        
        Args:
            gateway_id: Gateway that captured the counter
        """
        self._captured_slot_counters[gateway_id.value] = datetime.now()
    
    def gateway_slot_counter_observed(self, gateway_id: GatewayID, slot_counter: SlotCounter) -> None:
        """Gateway slot counter was observed (in RECEIVE_RESPONSE).
        
        Args:
            gateway_id: Gateway ID
            slot_counter: Observed slot counter value
        """
        # Get the capture time
        capture_time = self._captured_slot_counters.pop(gateway_id.value, None)
        if capture_time is None:
            return
        
        # Update or create slot clock
        if gateway_id.value in self._slot_clocks:
            try:
                self._slot_clocks[gateway_id.value].set(slot_counter, capture_time)
            except Exception:
                pass
        else:
            try:
                clock = SlotClock(slot_counter, capture_time)
                self._slot_clocks[gateway_id.value] = clock
            except Exception:
                pass
    
    def packet_received(
        self,
        gateway_id: GatewayID,
        header: 'ReceivedPacketHeader',
        data: bytes
    ) -> None:
        """PV packet was received (no-op at observer level)."""
        pass
    
    def command_executed(
        self,
        gateway_id: GatewayID,
        request: Tuple[int, bytes],
        response: Tuple[int, bytes],
    ) -> None:
        """Command was executed (no-op at observer level)."""
        pass
    
    # Implement application.Sink
    
    def string_request(self, gateway_id: GatewayID, pv_node_id: NodeID, request: str) -> None:
        """String request was sent."""
        pass
    
    def string_response(self, gateway_id: GatewayID, pv_node_id: NodeID, response: str) -> None:
        """String response was received."""
        pass
    
    def node_table_page(
        self,
        gateway_id: GatewayID,
        start_address: NodeAddress,
        nodes: list
    ) -> None:
        """Node table page was received — accumulate into node table.
        
        Args:
            gateway_id: Gateway ID
            start_address: Starting address of this page
            nodes: List of (NodeAddress, LongAddress) tuples
        """
        gw_val = gateway_id.value
        
        # Get or create builder for this gateway
        if gw_val not in self._node_table_builders:
            self._node_table_builders[gw_val] = {}
        
        builder = self._node_table_builders[gw_val]
        
        # Add entries from this page
        for node_addr, long_addr in nodes:
            node_id = node_addr.to_node_id()
            if node_id is not None:
                builder[node_id.value] = long_addr
        
        # If this page is empty or incomplete, the table is complete
        if len(nodes) == 0:
            # Build final node table
            table = NodeTable()
            for nid_val, addr in builder.items():
                table.set(NodeID(nid_val), addr)
            
            self._persistent_state.gateway_node_tables[gw_val] = table
            del self._node_table_builders[gw_val]
            self.write_persistent_state()
    
    def topology_report(
        self,
        gateway_id: GatewayID,
        pv_node_id: NodeID,
        topology_report: bytes
    ) -> None:
        """Topology report was received."""
        pass
    
    def power_report(
        self,
        gateway_id: GatewayID,
        pv_node_id: NodeID,
        power_report: PowerReport
    ) -> None:
        """Power report was received - generate JSON event.
        
        Args:
            gateway_id: Gateway that received the report
            pv_node_id: PV node that sent the report
            power_report: Parsed power report
        """
        # Get slot clock for this gateway
        slot_clock = self._slot_clocks.get(gateway_id.value)
        if slot_clock is None:
            print(
                f"Warning: Discarding power report from gateway {gateway_id} "
                f"due to missing slot clock",
                file=sys.stderr
            )
            return
        
        # Create PowerReportEvent
        try:
            event_payload = PowerReportEvent.from_power_report(
                gateway_id,
                pv_node_id,
                slot_clock,
                power_report
            )
        except Exception as e:
            print(
                f"Warning: Discarding power report from gateway {gateway_id} "
                f"due to invalid slot counter: {e}",
                file=sys.stderr
            )
            return
        
        # Wrap in Event envelope
        event = Event.power_report(event_payload)
        
        # Output as JSON
        print(json.dumps(event.to_dict()), flush=True)
